cli-files/go-cat/cmd/root.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright © 2022 NAME HERE <EMAIL ADDRESS>
+*/
+package cmd
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+var (
+	ErrNoPath = errors.New("Please provide a file name or path to a file")
+	ErrDir    = errors.New("Please provide a file name or path to a file, cannot read a directory")
+)
+
+var rootCmd = &cobra.Command{
+	Use:   "go-cat",
+	Short: "go-cat allows you to see contents of a file",
+	Long:  `go-cat allows you to see contents of a file, just like the cat command`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) == 0 {
+			return ErrNoPath
+		}
+
+		for _, path := range args {
+			err := PrintFileContents(cmd.OutOrStdout(), path)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+
+	},
+}
+
+func PrintFileContents(Writer io.Writer, path string) error {
+	file, err := os.Open(path)
+	defer file.Close()
+	if err != nil {
+		return err
+	}
+
+	stat, err := file.Stat()
+	if err != nil {
+		return err
+	}
+
+	if stat.IsDir() {
+		return ErrDir
+	}
+
+	scanner := bufio.NewScanner(file)
```

Comment:
So I have updated the `fmt.Print `to print out the bytes as string. But am I correct to assume this is still converting bytes to string in memory? 

I have read a bit about the conversion and while it's definitely happening in memory for `string->bytes` conversion it's not very clear what how it's converted the other way round.

Reply:
The key here is to avoid thinking about `string`s at all - strings are really useful when we're dealing with text, but we're not necessarily here. Have a look at the documentation of [`io.Copy`](https://pkg.go.dev/io#Copy) and see if you can see how it may be useful, without once using the type `string` :)

Reply:
that's neat and quite simple, thanks!